src/rocq_elpi_utils.ml

```diff
@@ -655,92 +668,98 @@ let detype ?(keepunivs = false) env sigma t =
       let n, typ = get_GLambda_name_tgt typ in
       let aliastyp = if List.for_all (Names.Name.equal Anonymous) nl then None else Some (CAst.make (ci.ci_ind, nl)) in
-      (n, aliastyp, Some typ)
+      holes, (n, aliastyp, Some typ)
     in
     let constructs = Array.init (Array.length bl) (fun i -> (ci.ci_ind, i + 1)) in
-      let constagsl = get_cstr_tags (snd env) ci bl in
-      let eqnl = detype_eqns env constructs constagsl (ci, univs, params, bl) in
-      DAst.make @@ GCases (RegularStyle, pred, [ (tomatch, (alias, aliastyp)) ], eqnl)
-  and detype_eqns env constructs consnargsl bl =
+    let constagsl = get_cstr_tags (snd env) ci bl in
+    let holes, eqnl = detype_eqns env holes constructs constagsl (ci, univs, params, bl) in
+    holes, DAst.make @@ GCases (RegularStyle, pred, [ (tomatch, (alias, aliastyp)) ], eqnl)
+  and detype_eqns env holes constructs consnargsl bl =
     let ci, u, pms, bl = bl in
-    CArray.to_list (CArray.map3 (detype_eqn env u pms) constructs consnargsl bl)
-  and detype_eqn env u pms constr construct_nargs br =
+    let holes, l = CArray.fold_left_map_i (fun i holes c -> detype_eqn env holes u pms c consnargsl.(i) bl.(i)) holes constructs in
+    holes, CArray.to_list l
+  and detype_eqn env holes u pms constr construct_nargs br =
     let ctx, body = RobustExpand.branch (snd env) sigma constr u pms br in
     let branch = EConstr.it_mkLambda_or_LetIn body ctx in
     let make_pat decl env b ids =
       let env, na = push_rel decl env b in
       let ids = match Context.binder_name na with Names.Name.Name x -> Names.Id.Set.add x ids | _ -> ids in
       (DAst.make (PatVar (Context.binder_name na)), env, ids)
     in
-    let rec buildrec ids patlist env n b =
+    let rec buildrec ids patlist env holes n b =
       if Int.equal n 0 then
-        CAst.make
-        @@ (Names.Id.Set.elements ids, [ DAst.make @@ PatCstr (constr, List.rev patlist, Anonymous) ], aux env b)
+        let holes, b = aux env holes b in
+        holes, CAst.make
+          @@ (Names.Id.Set.elements ids, [ DAst.make @@ PatCstr (constr, List.rev patlist, Anonymous) ], b)
       else
         match EConstr.kind sigma b with
         | Lambda (x, t, b) ->
             let pat, env, new_ids = make_pat (LocalAssum (x, t)) env b ids in
-            buildrec new_ids (pat :: patlist) env (pred n) b
+            buildrec new_ids (pat :: patlist) env holes (pred n) b
         | LetIn (x, b, t, b') ->
             let pat, env, new_ids = make_pat (LocalDef (x, b, t)) env b' ids in
-            buildrec new_ids (pat :: patlist) env (pred n) b'
+            buildrec new_ids (pat :: patlist) env holes (pred n) b'
         | _ -> assert false
     in
-    buildrec Names.Id.Set.empty [] env (List.length ctx) branch
+    buildrec Names.Id.Set.empty [] env holes (List.length ctx) branch
   in
 
-  let x = aux (names_of_env env, env) t in
-  x
+  let x = aux (names_of_env env, env) Evar.Set.empty t in
```

Comment:
Note that `detype` takes a sigma. The set of evar names may be non-empty (it is in sigma) and you should start from that, or better avoid collisions. Maybe it is a set of names and not a set of evar number you have to carry.